cpp/include/cudf/io_readers.hpp

```diff
@@ -101,4 +101,139 @@ class JsonReader {
+  gdf_size_type header = 0;                 ///< Row of the header data, zero based counting; Default is zero.
+
+  std::vector<std::string> names;           ///< Ordered List of column names; Empty by default.
+  std::vector<std::string> dtype;           ///< Ordered List of data types; Empty by default.
+
+  std::vector<int> use_cols_indexes;        ///< Indexes of columns to be processed and returned; Empty by default - process all columns.
+  std::vector<std::string> use_cols_names;  ///< Names of columns to be processed and returned; Empty by default - process all columns.
+
+  std::vector<std::string> true_values;     ///< List of values to recognize as boolean True; Empty by default.
+  std::vector<std::string> false_values;    ///< List of values to recognize as boolean False; Empty by default.
+  std::vector<std::string> na_values;       /**< Array of strings that should be considered as NA. By default the following values are interpreted as NA: 
+                                            '', '#N/A', '#N/A N/A', '#NA', '-1.#IND', '-1.#QNAN', '-NaN', '-nan', '1.#IND', '1.#QNAN', 'N/A', 'NA', 'NULL',
+                                            'NaN', 'n/a', 'nan', 'null'. */
+  bool          keep_default_na = true;     ///< Keep the default NA values; true by default.
+  bool          na_filter = true;           ///< Detect missing values (empty strings and the values in na_values); true by default. Passing false can improve performance.
+
+  std::string   prefix;                     ///< If there is no header or names, prepend this to the column ID as the name; Default value is an empty string.
+  bool          mangle_dupe_cols = true;    ///< If true, duplicate columns get a suffix. If false, data will be overwritten if there are columns with duplicate names; true by default.
+
+  char          quotechar = '\"';           ///< Define the character used to denote start and end of a quoted item; default is '\"'.
+  gdf_csv_quote_style quoting = QUOTE_MINIMAL; ///< Defines reader's quoting behavior; default is QUOTE_MINIMAL.
+  bool          doublequote = true;         ///< Indicates whether to interpret two consecutive quotechar inside a field as a single quotechar; true by default.
+
+  csv_reader_args() = default;
+};
+
+/**---------------------------------------------------------------------------*
+ * @brief Class used to parse Json input and convert it into gdf columns.
+ *
+ *---------------------------------------------------------------------------**/
+class CsvReader {
+private:
+  class Impl;
+  std::unique_ptr<Impl> impl_;
+
+public:
+  /**---------------------------------------------------------------------------*
+   * @brief CsvReader defult constructor;
+   * 
+   * Calling read_XYZ() on a default-constructed object yields an empty table.
+   *---------------------------------------------------------------------------**/
+  CsvReader() noexcept;
+
+  CsvReader(CsvReader const &rhs);
```

Comment:
Generated Cython code is the only place where this is used:
```
CsvReader reader = Csvreader(args)
```
gets translated into
```
CsvReader reader();
reader = Csvreader(args);
```
This requires both the default constructor and the assignment operator. Because of this, I defined all special member functions, following the rule of five.

Reply:
Why not just do `CsvReader reader(args)`? I'm more of a rule of zero fan myself :)

Reply:
In fact, since the only member of `CsvReader` is `std::unique_ptr<Impl> impl_`, then you can already do the rule of zero. There's no reason to do anything other than use the defaults since the only member already has all the constructors/operators that you need.

Reply:
@jrhemstad  Unfortunately I need a default constructor for Cython. The only way to work around this is to declare a pointer in the Cython layer, instead of an object. With this, I should be able to remove all special functions. Let me know if this is preferred.

Reply:
Default constructor is just fine. Rule of 0 states "classes should not have custom destructors, copy/move constructors or copy/move assignment operators."